[Inductor max autotune] Multithreaded Precompilation

kadeng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119010
* #119009
* #119008
* #119007
* #119006
* #119004
* __->__ #119386

When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

kadeng committed:
[Inductor max autotune] Multithreaded Precompilation

When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

[ghstack-poisoned]

kadeng commented:
This is a recreated already accepted PR https://github.com/pytorch/pytorch/pull/119003

kadeng commented:
@pytorchbot merge

kadeng commented:
@pytorchbot merge

Skylion007 commented on torch/_inductor/autotune_process.py:
```diff
@@ -556,6 +565,13 @@ def __init__(
+    def precompile(self):
+        # Prepopulate CUDACodeCache
+        # may happen in separate Threadpool
+        log.debug("Precompiling %s", str(self))
```
```suggestion
        log.debug("Precompiling %s", self)
```
should call str o self I think?

Skylion007 commented on torch/_inductor/autotune_process.py:
```diff
@@ -556,6 +565,13 @@ def __init__(
+        # may happen in separate Threadpool
+        log.debug("Precompiling %s", str(self))
+        CUDACodeCache.load(self.source_code, "so")
+        log.debug("Done precompiling %s", str(self))
```
```suggestion
        log.debug("Done precompiling %s", self)
```

Skylion007 commented on torch/_inductor/codecache.py:
```diff
@@ -744,7 +745,7 @@ def load(
         compile the graph and save it to the cache.
         """
-        from filelock import FileLock
+        from filelock import FileLock  # type: ignore[import-not-found]
```
We just need to add `types-filelock` to our lintrunner install.

Skylion007 replied:
Opened a PR: https://github.com/pytorch/pytorch/pull/119390

kadeng replied:
thanks for the comments, I made some small corresponding changes..

kadeng committed:
Update on "[Inductor max autotune] Multithreaded Precompilation"


When using the Cutlass backend, the compilation
of CUDA source files can totally dominate the runtime required for the benchmarking done
as part of Autotuning.

This change adds a multithreaded precompilation phase, which serves to pre-populate the compilation cache ( both in-memory, and a
possible on-disk sccache ).

Also it ensures that no unneccessary compilation
and benchmarking steps are performed, which was peviously the case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

kadeng commented:
@pytorchbot merge

Skylion007 commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -298,14 +298,18 @@ def __init__(
         make_kernel_render: Callable[[CUDATemplateBuffer, Optional[List[IRNode]]], str],
         bmreq: CUDABenchmarkRequest,
-        template: "CUDATemplate",
+        template: "CUDATemplate",  # type: ignore[name-defined]
```
Huh, this type ignore shouldn't be necessary...

kadeng replied:
I add these type ignores only on linter errors when I run lintrunner locally. I also have the impression that these are a bit inconsistent, e.g. sometimes I had linter warnings about code that was just fine the day before.
